File: WoofAndroid/presentation/profile/src/main/kotlin/com/perrystreet/woof/presentation/profile/ui/ProfilePagerScreen.kt
package com.perrystreet.woof.presentation.profile.ui

import androidx.compose.runtime.Composable
import androidx.compose.ui.tooling.preview.PreviewParameter
import com.perrystreet.woof.designsystem.atomic.organisms.emptystate.OrgErrorState
import com.perrystreet.woof.designsystem.atomic.templates.TemplatePager
import com.perrystreet.woof.designsystem.preview.PreviewDevices
import com.perrystreet.woof.designsystem.preview.ThemeProvider
import com.perrystreet.woof.designsystem.preview.ThemedScreenPreview
import com.perrystreet.woof.designsystem.theme.ITheme
import com.perrystreet.woof.presentation.profile.ui.preview.ProfilePreviewData
import com.perrystreet.woof.presentation.profile.uimodel.ProfilePageUIModel
import com.perrystreet.woof.presentation.profile.viewmodel.ProfilePagerViewModel
import com.perrystreet.woof.utils.guard

@Composable
fun ProfilePagerScreen(
    state: ProfilePagerViewModel.State,
    pageContent: @Composable (page: ProfilePageUIModel) -> Unit,
) {
    guard(state.pages.isNotEmpty()) { return }

    TemplatePager(
        pageCount = state.pages.size,
        initialPage = state.initialPage,
        key = { page -> state.pages[page].id },
    ) { page ->
        pageContent(state.pages[page])
    }
}

@PreviewDevices
@Composable
private fun ProfilePagerScreenPreview(@PreviewParameter(ThemeProvider::class) theme: ITheme) {
    ThemedScreenPreview(theme = theme) {
        ProfilePagerScreen(
            state = ProfilePagerViewModel.State(pages = listOf(ProfilePreviewData.page()), initialPage = 0),
        ) { page ->
            OrgErrorState(title = page.name, message = page.photoUrl, actionText = "", onActionTap = {})
        }
    }
}

File: WoofIOS/SwiftPackages/Presentation/Profile/Sources/PresentationProfile/UI/ProfilePagerScreen.swift
import DesignSystem
import PresentationCommon
import SwiftUI

public struct ProfilePagerScreen<PageContent: View>: View {
    private let state: ProfilePagerViewModel.State
    private let pageContent: (ProfilePageUIModel) -> PageContent

    public init(state: ProfilePagerViewModel.State, @ViewBuilder pageContent: @escaping (ProfilePageUIModel) -> PageContent) {
        self.state = state
        self.pageContent = pageContent
    }

    public var body: some View {
        viewGuard(!state.pages.isEmpty) {
            TemplatePager(
                pageCount: state.pages.count,
                initialPage: state.initialPage,
                key: { page in state.pages[page].id }
            ) { page in
                pageContent(state.pages[page])
            }
        }
    }
}

#Preview("Pager") {
    ThemedScreenPreview(theme: WoofTheme.light()) {
        ProfilePagerScreen(state: ProfilePagerViewModel.State(pages: [ProfilePreviewData.page()], initialPage: 0)) { page in
            OrgErrorState(title: page.name, message: page.photoUrl, actionText: "", onActionTap: {})
        }
    }
}
